app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java

```diff
@@ -598,149 +654,136 @@ public int getFilePosition(final File file) {
-                                final File parent = file.getParentFile();
-                                if (parent != null) {
-                                    final File remap = new File(parent.getAbsolutePath(), "appdata-public (" + file.getName() + ")");
-                                    virtualMapping.put(remap, new File(absExt));
-                                    newData.add(remap);
-                                }
-                            }
-                        }
-                    }
-                }
+    // This function is not called on the main thread, so post to the UI thread
+    private synchronized void _loadFolder(final @NonNull File folder, final @Nullable File toShow) {
 
-                // Don't sort recent items - use the default order
-                if (!folder.equals(VIRTUAL_STORAGE_RECENTS)) {
-                    GsFileUtils.sortFiles(newData, _dopt.sortByType, _dopt.sortFolderFirst, _dopt.sortReverse);
+        final boolean folderChanged = !folder.equals(_currentFolder);
+
+        final List<File> newData = new ArrayList<>();
+
+        if (folder.equals(VIRTUAL_STORAGE_RECENTS)) {
+            newData.addAll(_dopt.recentFiles);
+        } else if (folder.equals(VIRTUAL_STORAGE_POPULAR)) {
+            newData.addAll(_dopt.popularFiles);
+        } else if (folder.equals(VIRTUAL_STORAGE_FAVOURITE)) {
+            newData.addAll(_dopt.favouriteFiles);
+        } else if (folder.isDirectory()) {
+            GsCollectionUtils.addAll(newData, folder.listFiles(GsFileBrowserListAdapter.this));
+        }
+
+        if (folder.equals(VIRTUAL_STORAGE_ROOT)) {
+            newData.addAll(_virtualMapping.keySet());
+        }
+
+        // Add all emulated folders under /storage/emulated
+        if (VIRTUAL_STORAGE_EMULATED.equals(folder)) {
+            newData.add(new File(folder, "0"));
+            for (int i = 1; i < 10; i++) {
+                final File f = new File(folder, String.valueOf(i));
+                if (GsFileUtils.canCreate(f)) {
+                    newData.add(f);
                 }
+            }
+        }
+
+        if (folder.getAbsolutePath().equals("/")) {
+            newData.add(new File(folder, VIRTUAL_STORAGE_ROOT.getName()));
+        }
+
+        GsCollectionUtils.deduplicate(newData);
 
-                if (canGoUp(folder)) {
-                    newData.add(0, folder.equals(new File("/storage/emulated/0")) ? new File("/storage/emulated") : folder.getParentFile());
+        // Don't sort recent items - use the default order
+        if (!folder.equals(VIRTUAL_STORAGE_RECENTS)) {
+            GsFileUtils.sortFiles(newData, _dopt.sortByType, _dopt.sortFolderFirst, _dopt.sortReverse);
+        }
+
+        // Testing if modtimes have changed (modtimes generally only increase)
+        final long modSum = GsCollectionUtils.accumulate(newData, (f, s) -> s + f.lastModified(), 0L);
+        final boolean modSumChanged = modSum != _prevModSum;
```

Comment:
@guanglinn See modTime based update

Reply:
Your solution based on the sum of modTime is better, it is more universal. My approach is just a simple solution.